src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,514 @@
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+            if ((uint)arrayIndex > (uint)array.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+            }
+            int count = Count;
+            if (array.Length - arrayIndex < count)
+            {
+                throw new ArgumentException("Destination array is not long enough to copy all the items in the collection. Check array index and length.");
+            }
+
+            Debug.Assert(Count <= _entries.Length);
+            Entry[] entries = _entries;
+            for (int i = 0; i < count; ++i)
+            {
+                Entry entry = entries[i];
+                array[i + arrayIndex] = new KeyValuePair<TKey, TValue>(entry.Key, entry.Value);
+            }
+        }
+
+        bool ICollection<KeyValuePair<TKey, TValue>>.Remove(KeyValuePair<TKey, TValue> item)
+        {
+            int index = IndexOf(item.Key);
+            if (index >= 0 && EqualityComparer<TValue>.Default.Equals(_entries[index].Value, item.Value))
+            {
+                RemoveAt(index);
+                return true;
+            }
+            return false;
+        }
+
+        int IList<KeyValuePair<TKey, TValue>>.IndexOf(KeyValuePair<TKey, TValue> item)
+        {
+            int index = IndexOf(item.Key);
+            if (index >= 0 && !EqualityComparer<TValue>.Default.Equals(_entries[index].Value, item.Value))
+            {
+                index = -1;
+            }
+            return index;
+        }
+
+        void IList<KeyValuePair<TKey, TValue>>.Insert(int index, KeyValuePair<TKey, TValue> item) => Insert(index, item.Key, item.Value);
+        #endregion
+
+        private void Resize(int newSize)
+        {
+            int[] newBuckets = new int[newSize];
+            Entry[] newEntries = new Entry[newSize];
+
+            int count = Count;
+            Array.Copy(_entries, newEntries, count);
+            for (int i = 0; i < count; ++i)
+            {
+                ref Entry entry = ref newEntries[i];
+                ref int next = ref newBuckets[entry.HashCode % newSize];
+                entry.Next = next - 1;
+                next = i + 1;
+            }
+
+            _buckets = newBuckets;
+            _entries = newEntries;
+        }
+
+        private int IndexOf(TKey key, out int hashCode)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            IEqualityComparer<TKey> comparer = _comparer;
+            hashCode = (comparer?.GetHashCode(key) ?? key.GetHashCode()) & 0x7FFFFFFF;
+            int[] buckets = _buckets;
+            int index = -1;
+            if (buckets.Length > 0)
+            {
+                int bucket = hashCode % buckets.Length;
+                index = buckets[bucket] - 1;
```

Comment:
Consider adding GetBucketIndex and GetEntryIndex (eg like [this](https://github.com/dotnet/corefxlab/pull/2458/files#diff-6f44e9ce636799efd5c8b26382f2248bR39) and [this](https://github.com/dotnet/corefxlab/pull/2458/files#diff-6f44e9ce636799efd5c8b26382f2248bR42)). That hides a lot of the +1 and -1 making the code easier to read, and they will certainly be inlined.

Reply:
It doesn't seem worth it as it would only be used once here as the other place I do this I use a ref variable.